File: web/managed-worker-service/agent_system_metrics.py
"""Optional psutil-based CPU/RAM sampling for eltPulse run telemetry."""

from __future__ import annotations

import asyncio
import os
from typing import Any, Callable, Awaitable

PatchFn = Callable[[dict[str, Any]], Awaitable[None]]


def system_metrics_enabled() -> bool:
    raw = (os.environ.get("ELTPULSE_SYSTEM_METRICS") or "").strip().lower()
    if raw in ("0", "false", "no"):
        return False
    return True


def system_metrics_interval_sec() -> float:
    raw = (os.environ.get("ELTPULSE_SYSTEM_METRICS_INTERVAL_MS") or "").strip()
    try:
        ms = int(raw)
        if ms >= 5000:
            return min(ms, 120_000) / 1000.0
    except ValueError:
        pass
    return 20.0


def read_system_metrics(pid: int | None = None) -> dict[str, float] | None:
    """Read CPU % and RSS MB for a process (defaults to current worker)."""
    try:
        import psutil  # type: ignore[import-not-found]
    except ImportError:
        return None
    try:
        proc = psutil.Process(pid) if pid is not None else psutil.Process()
        mem_mb = round(proc.memory_info().rss / (1024 * 1024), 1)
        cpu = round(proc.cpu_percent(interval=0.1), 1)
        out: dict[str, float] = {"memoryMb": mem_mb}
        if cpu >= 0:
            out["cpuPercent"] = cpu
        return out
    except Exception:
        return None


def telemetry_patch_with_system(system: dict[str, float] | None) -> dict[str, Any]:
    if not system:
        return {}
    return {
        "telemetrySummary": {"system": system},
        "appendTelemetrySample": {"system": system},
    }


async def sample_system_metrics_loop(
    patch: PatchFn,
    stop: asyncio.Event,
    *,
    pid: int | None = None,
) -> None:
    """Background task: PATCH system metrics every N seconds until stop is set."""
    if not system_metrics_enabled():
        return
    interval = system_metrics_interval_sec()
    while not stop.is_set():
        system = read_system_metrics(pid)
        if system:
            try:
                body: dict[str, Any] = {"status": "running", **telemetry_patch_with_system(system)}
                await patch(body)
            except Exception:
                pass
        try:
            await asyncio.wait_for(stop.wait(), timeout=interval)
        except asyncio.TimeoutError:
            continue

File: web/lib/elt/agent-system-metrics.ts
/**
 * Optional worker CPU/RAM sampling for run telemetry (`telemetrySummary.system`).
 * Works in Node gateways and managed executors without extra deps.
 *
 * Enable: ELTPULSE_SYSTEM_METRICS=1 (default on). Disable: ELTPULSE_SYSTEM_METRICS=0
 */

export type SystemMetrics = {
  cpuPercent?: number;
  memoryMb?: number;
};

export function systemMetricsEnabled(): boolean {
  const raw = process.env.ELTPULSE_SYSTEM_METRICS?.trim().toLowerCase();
  if (raw === "0" || raw === "false" || raw === "no") return false;
  return true;
}

export function systemMetricsIntervalMs(): number {
  const n = Number(process.env.ELTPULSE_SYSTEM_METRICS_INTERVAL_MS ?? "");
  if (Number.isFinite(n) && n >= 5_000) return Math.min(n, 120_000);
  return 20_000;
}

let lastCpuUsage: NodeJS.CpuUsage | null = null;
let lastCpuAt = 0;

/** Best-effort process metrics for the current Node worker (no native addons). */
export function readProcessSystemMetrics(): SystemMetrics | null {
  const mem = process.memoryUsage();
  const memoryMb = Math.round((mem.rss / (1024 * 1024)) * 10) / 10;

  const now = Date.now();
  let cpuPercent: number | undefined;
  const usage = process.cpuUsage();
  if (lastCpuUsage && lastCpuAt > 0) {
    const elapsedMs = now - lastCpuAt;
    if (elapsedMs > 0) {
      const userDelta = usage.user - lastCpuUsage.user;
      const sysDelta = usage.system - lastCpuUsage.system;
      const cpuMs = (userDelta + sysDelta) / 1000;
      cpuPercent = Math.min(100, Math.round((cpuMs / elapsedMs) * 1000) / 10);
    }
  }
  lastCpuUsage = usage;
  lastCpuAt = now;

  return {
    memoryMb,
    ...(cpuPercent !== undefined ? { cpuPercent } : {}),
  };
}

export function telemetryPatchWithSystem(metrics: SystemMetrics | null): {
  telemetrySummary?: { system: SystemMetrics };
  appendTelemetrySample?: { system: SystemMetrics };
} {
  if (!metrics) return {};
  return {
    telemetrySummary: { system: metrics },
    appendTelemetrySample: { system: metrics },
  };
}

/** Poll until `stop()` is called; invokes `patch` with telemetry fragments. */
export function startSystemMetricsSampler(patch: (body: Record<string, unknown>) => Promise<void>): {
  stop: () => void;
} {
  if (!systemMetricsEnabled()) {
    return { stop: () => undefined };
  }

  let stopped = false;
  const intervalMs = systemMetricsIntervalMs();

  const tick = async () => {
    if (stopped) return;
    const system = readProcessSystemMetrics();
    if (!system) return;
    try {
      await patch({
        status: "running",
        ...telemetryPatchWithSystem(system),
      });
    } catch {
      /* non-fatal */
    }
  };

  void tick();
  const handle = setInterval(() => void tick(), intervalMs);

  return {
    stop: () => {
      stopped = true;
      clearInterval(handle);
    },
  };
}
